Create a Getting Started Guide for Database Monitoring (#11727)
DataDog/documentation: content/en/getting_started/database_monitoring/_index.md

---
title: Getting Started with Database Monitoring
kind: documentation
further_reading:
    - link: '/database_monitoring/'
      tag: 'Documentation'
      text: 'Database Monitoring'
    - link: '/database_monitoring/troubleshooting/'
      tag: 'Documentation'
      text: 'Troubleshooting'
    - link: '/blog/database-performance-monitoring-datadog/'
      tag: 'Blog'
      text: 'Database performance monitoring with Datadog'
---
## Overview

Datadog Database Monitoring allows you to easily understand the health and performance of your databases, and quickly determine the root cause of any problems.

In one place, you can view:

* Host-level metrics
* Explain plans
* Historical query performance metrics

> In this guide, you will learn how to set up Datadog Database Monitoring on an example PostgreSQL database. You will identify an expensive query, troubleshoot a slow query, and quickly view changes in query volume.

Reviewer: ```suggestion
Work through this guide to set up Datadog Database Monitoring on an example PostgreSQL database. Next, identify an expensive query, troubleshoot a slow query, and create a dashboard to view changes in query volume.
```

## Setup

### Prerequisites

> Before getting started, you need a Datadog account.

Reviewer 2: ```suggestion
Before getting started, you need a [Datadog account][6].
```

To run the example application, you need a machine with [GNU Make][1] and [Docker][2]. Have your Datadog [API key][3] available.

### Install the example application

The example application starts up the Datadog Agent and a PostgreSQL database in a Docker container. While the application runs, the Agent sends database metrics to Datadog. You can view the data from the application in Datadog Database Monitoring.

Follow these instructions to install the example application on MacOS or Linux.

1. Download the [archive][4] containing the example application.

2. Extract the archive to a location on disk:
```
tar -xvf database-management.tgz
```

3. Change to the `database-management` directory in the extracted archive:
```
cd database-management
```

4. Set the environment variable `DD_API_KEY` to your Datadog API key:
```
export DD_API_KEY=<API_KEY>
```

5. Start the application:
```
make postgres
```

> The command will continue to run until you stop it by pressing Ctrl + C.

Reviewer: ```suggestion
The command continues to run until you stop it by pressing Ctrl + C.
```

## Identify an expensive query

Which query consumes the most database time? To find out, use the Query Metrics view.

> 1. Navigate to Database Monitoring by clicking **APM > Databases** in the UI. You will see the Query Metrics view.

Reviewer: ```suggestion
1. Navigate to Database Monitoring by clicking **APM > Databases** in the UI. It opens to the Query Metrics view.
```

2. Sort the Normalized Query table by **Percent time** to see the query that the database spends the most time executing.

> The query that consumes the most database time will appear on the first line.

Reviewer: ```suggestion
The query that consumes the most database time appears on the first line.
```

## View changes in query volume

To understand the health and performance of your databases at a glance, add Datadog Database Monitoring metrics to a dashboard.

For example, you can see the absolute change in query volume in the past hour by adding a **Change** widget to track a query count metric.

1. Select **Dashboards > New Dashboard** in the UI.

2. Enter a name for your dashboard. Click the **New Dashboard** button to go to your new dashboard.

2. To add content to your dashboard, click **Add Widgets**.

3. In the widget carousel, select the **Change** widget.

4. Select `postgresql.queries.count` in the **Metric** dropdown. This metric counts the number of queries sent to a PostgreSQL database.

5. Select `host` in the **Break it down by** dropdown so the widget aggregates queries by host.

{{< img src="database_monitoring/dashboard_change_postgres.png" alt="Configure change widget for postgres queries metric" style="width:100%;">}}

7. Click the **Save** button. The dashboard shows your new widget.

{{< img src="database_monitoring/dashboard_change_widget.png" alt="Change widget showing query count" style="width:100%;">}}

> 

Reviewer: Here, let's wrap up by telling them that there is also an out-of-the-box dashboard for Postgres and MySQL that they can access through the Dashboards button on the DBM page, and which they can clone and modify to suit their needs.

## Further Reading

{{< partial name="whats-next/whats-next.html" >}}

[1]: https://www.gnu.org/software/make/
[2]: https://www.docker.com/
[3]: https://app.datadoghq.com/account/settings#api
[4]: /resources/examples/database-management.tgz

> [5]: https://app.datadoghq.com/databases

Reviewer 2: ```suggestion
[5]: https://app.datadoghq.com/databases
[6]: https://www.datadoghq.com/free-datadog-trial/
```

In one place, you can view:

* Host-level metrics
* Explain plans
* Historical query performance metrics

Work through this guide to set up Datadog Database Monitoring on an example PostgreSQL database. Next, identify an expensive query, troubleshoot a slow query, and create a dashboard to view changes in query volume.

## Setup

### Prerequisites

Before getting started, you need a [Datadog account][1].

To run the example application, you need a machine with [GNU Make][2] and [Docker][3]. Have your Datadog [API key][4] available.

### Install the example application

The example application starts up the Datadog Agent and a PostgreSQL database in a Docker container. While the application runs, the Agent sends database metrics to Datadog. You can view the data from the application in Datadog Database Monitoring.

Follow these instructions to install the example application on MacOS or Linux.

1. Download the [archive][5] containing the example application.

2. Extract the archive to a location on disk:
```
tar -xvf database-management.tgz
```

> 3. Change to the `database-management` directory in the extracted archive:

Reviewer 3: Where does `database-management` come from? This could be confusing since the name of the product is `database monitoring`.

Author: oops, thank you for catching this! `database-management` is a typo on my part.

```
cd database-management
```

4. Set the environment variable `DD_API_KEY` to your Datadog API key:
```
export DD_API_KEY=<API_KEY>
```

5. Start the application:
```

> make postgres

Reviewer 3: Should we allow them to also run `make mysql`

Author: I originally had separate tabs for the two targets, but then I tested the mysql one and it failed on my machine. I decided to take it out until I got a chance to talk to you about the problem.

Reviewer 3: 👍 sounds good, let me know what the issue is and i'll try to get it worked out

1. Select **Dashboards > New Dashboard** in the UI.

2. Enter a name for your dashboard. Click the **New Dashboard** button to go to your new dashboard.

2. To add content to your dashboard, click **Add Widgets**.

3. In the widget carousel, select the **Change** widget.

4. Select `postgresql.queries.count` in the **Metric** dropdown. This metric counts the number of queries sent to a PostgreSQL database.

5. Select `host` in the **Break it down by** dropdown so the widget aggregates queries by host.

{{< img src="database_monitoring/dashboard_change_postgres.png" alt="Configure change widget for postgres queries metric" style="width:100%;">}}

7. Click the **Save** button. The dashboard shows your new widget.

{{< img src="database_monitoring/dashboard_change_widget.png" alt="Change widget showing query count" style="width:100%;">}}

### View default dashboards

Observe current database activity, resource utilization, and more on out-of-the-box dashboards provided by Datadog Database Monitoring.

To access the dashboards:

1. Click **APM > Databases** in the UI to go to Database Monitoring.

> 2. Select the **Dashboards** tab.
> 
> 3. Choose one of the database options:
>    - _MySQL_
>    - _AWS RDS (MySQL)_
>    - _PostgreSQL_
>    - _AWS RDS (PostgreSQL)_

Reviewer: ```suggestion
2. Select the **Dashboards** tab and choose the dashboard you want to see.
```

> The dashboard you selected opens. You may clone or modify the dashboard to suit your needs.

Reviewer: ```suggestion
You can clone and modify out-of-the-box dashboards to suit your needs.
```
